mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -165,154 +217,176 @@ impl RingBuffer {
+        if self.metadata.file_pointers.read_end == write_index
+            && !self.metadata.can_read_from_wrap_around
+        {
             return Ok(VecDeque::new());
         }
 
         // If we are reading then we can reset wrap around flag
-        self.can_read_from_wrap_around = false;
+        self.metadata.can_read_from_wrap_around = false;
 
         let mut start = read_index;
         let mut vdata = VecDeque::new();
         for _ in 0..count {
-            let end = start + block_size;
-            let bytes = &self.mmap[start as usize..end as usize];
-            // this is unused memory
-            if bytes.iter().map(|&x| x as usize).sum::<usize>() == 0 {
-                break;
-            }
+            #[allow(clippy::cast_possible_truncation)]
+            let block = load_block_header(
+                &mut self.file,
+                start,
+                block_size as usize,
+                self.max_file_size,
+            )?;
 
-            let block = bincode::deserialize::<BlockHeaderWithHash>(bytes)?;
             // this means we read bytes that don't make a block, this is
             // a really bad state to be in as somehow the pointers don't
             // match to where data really is at.
-            if block.inner().hint() != BLOCK_HINT {
+            let BlockVersion::Version1(inner_block) = block.inner();
+            if inner_block.hint() != BLOCK_HINT {
+                self.file.seek(SeekFrom::Start(0)).unwrap();
+                let mut buf = vec![];
+                self.file.read_to_end(&mut buf).unwrap();
+
                 return Err(RingBufferError::Validate(BlockError::Hint).into());
             }
 
-            let inner_block = block.inner();
             let data_size = inner_block.data_size();
             let index = inner_block.write_index();
             start += block_size;
             let end = start + data_size;
-            let data = load_data(
-                &self.mmap,
-                start as usize,
+            #[allow(clippy::cast_possible_truncation)]
+            let publication = load_data(
+                &mut self.file,
+                start,
                 data_size as usize,
-                self.max_file_size as usize,
+                self.max_file_size,
             )?;
+
+            // Update start for the next block.
             start = end % self.max_file_size;
-            self.pointers.read_end = start;
+            self.metadata.file_pointers.read_end = start;
 
-            validate(&block, &bincode::serialize(&data)?)?;
-            let key = Key { offset: index };
+            // Validate the block and data. This should be data section in the mmap
+            // after a `BlockHeaderWithHash`.
+            validate(&block, &bincode::serialize(&publication)?)?;
 
-            vdata.push_back((key, data));
+            #[allow(clippy::cast_possible_truncation)]
+            let key = Key {
+                offset: index as u32,
```

Comment:
why u32?

Reply:
From previous fix pr, this is u32. This will be resolved as u64 again in my other integration pr